### PR TITLE
More Maven Property Resolving

### PR DESCRIPTION
**Pull request by mikeyoung85**
This will resolve properties that are hanging out inside of a string like `http://www.myhomepage.com/${project.artifactId}`. It will also recurse up to 5 times to resolve a property. Added some changes to the spec pom file to verify the resolved values.

### CODE REVIEW

File: lib/bibliothecary/parsers/maven.rb [hunk 1]
```diff
@@ -172,30 +174,53 @@ def self.extract_pom_info(xml, location, parent_properties = {})
...
-          elsif parent_prop
-            return parent_prop
-          elsif xml.locate(non_prop_name).first
-            # see if the value to look up is a field under the project
-            # examples are ${project.groupId} or ${project.version}
-            return xml.locate(non_prop_name).first.nodes.first
-          else
-            return value
-          end
+          return extract_property(xml, match[2], value, parent_properties)
```
Comment (kbarrette) [inline lib/bibliothecary/parsers/maven.rb]: If we only ever care about the second capture, we can ditch the first.

File: lib/bibliothecary/parsers/maven.rb [hunk 2]
```diff
@@ -12,6 +12,8 @@ class Maven
       # "|    \\--- com.google.guava:guava:23.5-jre (*)"
       GRADLE_DEP_REGEX = /(\+---|\\---){1}/
 
+      MAVEN_PROPERTY_REGEX = /^(.*)\$\{(.+)\}(.*)/
```
Comment (kbarrette) [inline lib/bibliothecary/parsers/maven.rb]: I think `/\$/{(.+?)\}/` would do here

File: lib/bibliothecary/parsers/maven.rb [hunk 3]
```diff
@@ -172,30 +174,53 @@ def self.extract_pom_info(xml, location, parent_properties = {})
...
         else
           return value
         end
       end
+
+      def self.replace_value_with_prop(original_value, property_value, property_name)
+        original_value.gsub("${#{property_name}}", property_value)
+      end
+
+      def self.extract_property(xml, property_name, value, parent_properties = {}, depth = 0)
```
Comment (kbarrette) [inline lib/bibliothecary/parsers/maven.rb]: I think if you break this into two methods it'll flow more easily, something like:
```ruby
      def self.extract_property(xml, property_name, value, parent_properties = {}, depth = 0)
        prop_value = property_value(xml, property_name, parent_properties)
        return value unless prop_value
        resolved_value = replace_value_with_prop(value, prop_value, property_name)

        # check to see if we just resolved to another property name
        match = resolved_value.match(MAVEN_PROPERTY_REGEX)
        if match && depth < MAX_DEPTH
          extract_property(xml, match[2], resolved_value, parent_properties, depth + 1)
        else
          resolved_value
        end
      end

      def self.property_value(xml, property_name, parent_properties)
        # the xml root is <project> so lookup the non property name in the xml
        # this converts ${project/group.id} -> ${group/id}
        prop_field = xml.respond_to?(:properties) && xml.properties.locate(property_name).first
        non_prop_name = property_name.gsub(".", "/").gsub("project/", "")

        if prop_field
          prop_field.nodes.first
        elsif parent_properties[property_name]
          parent_properties[property_name]
        elsif (non_prop = xml.locate(non_prop_name).first)
          # see if the value to look up is a field under the project
          # examples are ${project.groupId} or ${project.version}
          non_prop.nodes.first
        end

        # property not found
      end
```

File: lib/bibliothecary/parsers/maven.rb (hunk 2 above)
Comment (mikeyoung85) [reply lib/bibliothecary/parsers/maven.rb]: It wouldn't capture any properties in the middle of the string, which was one of the problems we were having.

File: lib/bibliothecary/parsers/maven.rb (hunk 3 above)
Comment (kbarrette) [reply lib/bibliothecary/parsers/maven.rb]: n.b. I did not test this

File: lib/bibliothecary/parsers/maven.rb (hunk 2 above)
Comment (kbarrette) [reply lib/bibliothecary/parsers/maven.rb]: That was due to the `^` marking "beginning of line"

**Pull request merged**